Blog post on Quarkus Agentic AI

mariofusco opened the pull request:
@cescoffier @geoand Please review.

geoand reviewed (approved):
Very nice!

geoand commented:
Just small nitpicking: The commit message is weird :)

mariofusco commented:
> Just small nitpicking: The commit message is weird :)

I messed up with git yesterday evening :) Fixed.

geoand commented:
🙏🏽

cescoffier commented on _posts/2025-02-18-agentic-ai-with-quarkus.adoc:
```diff
@@ -0,0 +1,178 @@
+quarkus.langchain4j.ollama.big-model.timeout=600s
+----
+
+it is possible to dare to implement an AI agent with more freedom to operate, in order to make it capable to cover a larger range of tasks. This https://github.com/mariofusco/quarkus-agentic-ai/blob/main/src/main/java/org/agenticai/searchastool/IntelligentAgent.java[intelligent agent] has been provided with only https://github.com/mariofusco/quarkus-agentic-ai/blob/main/src/main/java/org/agentica ...
```
```suggestion
Implementing an AI agent with more freedom to operate is possible. Thus, agents can cover a more extensive range of tasks. This https://github.com/mariofusco/quarkus-agentic-ai/blob/main/src/main/java/org/agenticai/searchastool/IntelligentAgent.java[intelligent agent] has been provided with only https://github.com/mariofusco/quarkus-agentic-ai/blob/main/src/main/java/org/agenticai/searchastool/WebSearchService.java[one tool] that can perform a web search on _DuckDuckGo_ and returns the result in plain text. The AI agent uses this tool to retrieve online information for everything it ignores and arrange that data together to provide an answer to a generic user question.

These more capable agents require two main ingredients:

1. A model able to reason and decide which tools to call with which parameter
2. A set of tools (toolbox in the Quarkus lingo) giving the LLM more data and capabilities
```

cescoffier commented on _posts/2025-02-18-agentic-ai-with-quarkus.adoc:
```diff
@@ -0,0 +1,178 @@
+
+it is possible to dare to implement an AI agent with more freedom to operate, in order to make it capable to cover a larger range of tasks. This https://github.com/mariofusco/quarkus-agentic-ai/blob/main/src/main/java/org/agenticai/searchastool/IntelligentAgent.java[intelligent agent] has been provided with only https://github.com/mariofusco/quarkus-agentic-ai/blob/main/src/main/java/org/agentica ...
+
+For instance we may want to ask `how many seconds would it take for a leopard at full speed to run through Pont des Arts` making a request to the https://github.com/mariofusco/quarkus-agentic-ai/blob/main/src/main/java/org/agenticai/searchastool/AgenticChatbotResource.java[rest endpoint] exposing this service like it follows
```
```suggestion
For instance, consider the following question: How many seconds would it take for a leopard at full speed to run through the Pont des Arts? Using this https://github.com/mariofusco/quarkus-agentic-ai/blob/main/src/main/java/org/agenticai/searchastool/AgenticChatbotResource.java[HTTP endpoint], it would be executed using:
```

cescoffier commented on _posts/2025-02-18-agentic-ai-with-quarkus.adoc:
```diff
@@ -0,0 +1,178 @@
+
+For instance we may want to ask `how many seconds would it take for a leopard at full speed to run through Pont des Arts` making a request to the https://github.com/mariofusco/quarkus-agentic-ai/blob/main/src/main/java/org/agenticai/searchastool/AgenticChatbotResource.java[rest endpoint] exposing this service like it follows
+
+http://localhost:8080/ask/how%20many%20seconds%20would%20it%20take%20for%20a%20leopard%20at%20full%20speed%20to%20run%20through%20Pont%20des%20Arts
```
```suggestion
```
> curl http://localhost:8080/ask/how%20many%20seconds%20would%20it%20take%20for%20a%20leopard%20at%20full%20speed%20to%20run%20through%20Pont%20des%20Arts
```
```

cescoffier commented on _posts/2025-02-18-agentic-ai-with-quarkus.adoc:
```diff
@@ -0,0 +1,178 @@
+
+http://localhost:8080/ask/how%20many%20seconds%20would%20it%20take%20for%20a%20leopard%20at%20full%20speed%20to%20run%20through%20Pont%20des%20Arts
+
+To reply to this question the intelligent agent invokes the web search tool twice, once to find the length of Pont des Arts and once to retrieve the speed of a leopard
```
```suggestion
To reply to this question,, the agent invokes the web search tool twice: once to find the length of Pont des Arts and once to retrieve a leopard's speed.
```

BTW, which Pont des Arts are you talking about?

cescoffier commented on _posts/2025-02-18-agentic-ai-with-quarkus.adoc:
```diff
@@ -0,0 +1,178 @@
+.An agent using an external web search tool
+image::ai-agent.png[]
+
+and then puts this information together generating an output like the following.
```
```suggestion
Then, the agent puts this information together and  generates an output like:
```

cescoffier commented on _posts/2025-02-18-agentic-ai-with-quarkus.adoc:
```diff
@@ -0,0 +1,178 @@
+
+=== Agents and Conversational AI
+
+The flexibility of AI agents can become even more relevant when used in services not intended to fulfil a single request, but needing to have a longer conversation with the user in order to achieve their goal. For instance agents can function as chatbots, enabling them to handle multiple users in parallel, each with their own independent conversation. This requires managing the state of each conv ...
```
```suggestion
The flexibility of AI agents can become even more relevant when used in services that are not intended to fulfill a single request but need to have a longer conversation with the user to achieve their goal. For instance, agents can function as chatbots, enabling them to handle multiple users in parallel, each with their own independent conversation. This requires managing the state of each conversation, often referred to as memories (the set of messages already exchanged with the LLM).
```

cescoffier commented on _posts/2025-02-18-agentic-ai-with-quarkus.adoc:
```diff
@@ -0,0 +1,178 @@
+
+The flexibility of AI agents can become even more relevant when used in services not intended to fulfil a single request, but needing to have a longer conversation with the user in order to achieve their goal. For instance agents can function as chatbots, enabling them to handle multiple users in parallel, each with their own independent conversation. This requires managing the state of each conv ...
+
+A https://github.com/mariofusco/quarkus-agentic-ai/blob/main/src/main/java/org/agenticai/restaurant/RestaurantAgent.java[chatbot of a restaurant booking system], designed to chat with customers and collect their data and requirements, represents an interesting practical application of this pattern. When the agent completes that information gathering process, the chatbot uses a https://github.com/ ...
```
```suggestion
A https://github.com/mariofusco/quarkus-agentic-ai/blob/main/src/main/java/org/agenticai/restaurant/RestaurantAgent.java[chatbot of a restaurant booking system], designed to chat with customers and collect their data and requirements, represents an interesting practical application of this pattern. When the agent completes that information-gathering process, the chatbot uses a https://github.com/mariofusco/quarkus-agentic-ai/blob/main/src/main/java/org/agenticai/restaurant/booking/BookingService.java[tool accessing the database] of existing reservations to both check if there is still a table available for the customers needs and to place a booking of that table if so. To assist the customer in deciding whether it is appropriate to eat outside or not, the agent can also reuse, as a second tool, the https://github.com/mariofusco/quarkus-agentic-ai/blob/main/src/main/java/org/agenticai/restaurant/weather/WeatherService.java[weather forecast service] implemented in one of the former examples, passing to it the geographic coordinates of the restaurant.
```

cescoffier commented on _posts/2025-02-18-agentic-ai-with-quarkus.adoc:
```diff
@@ -0,0 +1,178 @@
+.The restaurant chatbot agent
+image::restaurant-agent.png[]
+
+Once the customer provides all necessary details, the chatbot confirms the booking and presents a summary of the reservation. and the final booking is then stored in the database. It is possible to give a try to this chatbot accessing the URL:
```
```suggestion
Once the customer provides all necessary details, the chatbot confirms the booking and presents a reservation summary. The final booking is then stored in the database. It is possible to give this a try by accessing the URL:
```

cescoffier commented on _posts/2025-02-18-agentic-ai-with-quarkus.adoc:
```diff
@@ -0,0 +1,178 @@
+
+http://localhost:8080/restaurant.html
+
+and a typical example of user interaction could be something like this:
```
```suggestion
A typical example of user interaction could be something like this:
```

cescoffier commented on _posts/2025-02-18-agentic-ai-with-quarkus.adoc:
```diff
@@ -0,0 +1,178 @@
+and a typical example of user interaction could be something like this:
+
+.An example of interaction with the restaurant chatbot agent
+image::restaurant.png[]
```
We would need a conclusion pointing to the next steps:

- memory management across LLM calls 
- improving observability 
- dynamic tools (and tools registry)
- agentic pattern and MCP

cescoffier reviewed (changes requested):
The content is great. My main concern is the lack of code examples in the blog post itself, which requires users to navigate to GitHub. We would need to add snippets to improve understandability.

mariofusco commented on _posts/2025-02-18-agentic-ai-with-quarkus.adoc:
```diff
@@ -0,0 +1,178 @@
+== Agents
+
+Agents differ from the workflow patterns discussed so far because the control flow is entirely delegated to LLMs, that typically invoke tools to complete determinate subtask, instead of being implemented programmatically. These tools can be seen as stateless functions, opportunistically invoked by the LLM to achieve its goal and implemented in a variety of different technologies, like simple impe ...
+
```
In my opinion the tracings were only interesting to demonstrate the sequential vs. parallel workflows, is there anything else that could be relevant to demonstrate with the tracings of the execution of these agents?

cescoffier replied:
MY comment was actually more about the routing approach with 2 traces:

- prompt -> router -> medical
- prompt -> router -> technical

mariofusco commented:
@cescoffier I believe that I addressed all your suggestions, please give a look, especially to the conclusions because I don't know if they are enough or you want to explain those future works in more details.

With the introduction of the code parts (that I originally left out to keep the article more compact), we now have a quite long post, so I'm wondering if it could make sense to split it in 2 parts, also considering that it has a very natural splitting point, separating the workflow and the pure agentic sections.
